Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveM.cpp

```diff
@@ -0,0 +1,219 @@
+            Array4<Real> const& Hy_bias = H_biasfield[1]->array(mfi); // Hy_bias is the y component at |_y faces
+            Array4<Real> const& Hz_bias = H_biasfield[2]->array(mfi); // Hz_bias is the z component at |_z faces
+            Array4<Real> const& Bx = Bfield[0]->array(mfi); // Bx is the x component at |_x faces
+            Array4<Real> const& By = Bfield[1]->array(mfi); // By is the y component at |_y faces
+            Array4<Real> const& Bz = Bfield[2]->array(mfi); // Bz is the z component at |_z faces
+
+            // extract stencil coefficients
+            Real const * const AMREX_RESTRICT coefs_x = m_stencil_coefs_x.dataPtr();
+            int const n_coefs_x = m_stencil_coefs_x.size();
+            Real const * const AMREX_RESTRICT coefs_y = m_stencil_coefs_y.dataPtr();
+            int const n_coefs_y = m_stencil_coefs_y.size();
+            Real const * const AMREX_RESTRICT coefs_z = m_stencil_coefs_z.dataPtr();
+            int const n_coefs_z = m_stencil_coefs_z.size();
+
+            // extract tileboxes for which to loop
+            Box const& tbx = mfi.tilebox(Bfield[0]->ixType().toIntVect()); /* just define which grid type */
+            Box const& tby = mfi.tilebox(Bfield[1]->ixType().toIntVect());
+            Box const& tbz = mfi.tilebox(Bfield[2]->ixType().toIntVect());
+
+            // loop over cells and update fields
+            amrex::ParallelFor(tbx, tby, tbz,
+            [=] AMREX_GPU_DEVICE (int i, int j, int k){
+
+              // when working on M_xface(i,j,k, 0:2) we have direct access to M_xface(i,j,k,0:2) and Hx(i,j,k)
+              // Hy and Hz can be acquired by interpolation
+              // H_maxwell
+              Real Hx_xface = MacroscopicProperties::getH_Maxwell(i, j, k, 0, amrex::IntVect(1,0,0), amrex::IntVect(1,0,0), Bx, M_xface);
+              Real Hy_xface = MacroscopicProperties::getH_Maxwell(i, j, k, 1, amrex::IntVect(0,1,0), amrex::IntVect(1,0,0), By, M_xface);
+              Real Hz_xface = MacroscopicProperties::getH_Maxwell(i, j, k, 2, amrex::IntVect(0,0,1), amrex::IntVect(1,0,0), Bz, M_xface);
+              // H_bias
+              Real Hx_bias_xface = MacroscopicProperties::face_avg_to_face(i, j, k, 0, amrex::IntVect(1,0,0), amrex::IntVect(1,0,0), Hx_bias);
+              Real Hy_bias_xface = MacroscopicProperties::face_avg_to_face(i, j, k, 0, amrex::IntVect(0,1,0), amrex::IntVect(1,0,0), Hy_bias);
+              Real Hz_bias_xface = MacroscopicProperties::face_avg_to_face(i, j, k, 0, amrex::IntVect(0,0,1), amrex::IntVect(1,0,0), Hz_bias);
+              // H_eff = H_maxwell + H_bias + H_exchange + H_anisotropy ... (only the first two terms are considered here)
+              Real Hx_eff = Hx_bias_xface;
+              Real Hy_eff = Hy_bias_xface;
+              Real Hz_eff = Hz_bias_xface;
+
+              // Real Hx_eff = Hx_xface + Hx_bias_xface;
+              // Real Hy_eff = Hy_xface + Hy_bias_xface;
+              // Real Hz_eff = Hz_xface + Hz_bias_xface;
+
```

Comment:
Do you want to uncomment these lines to generalize the code now?
You could also save memory by saying:
Hx_xface += Hx_bias_xface;
and using Hx_xface instead of Hx_eff below.